### PR TITLE
fix: Inconsistencies in C++ guide

### PR DESCRIPTION
**Pull request by luator**
## Description
The C++ guide was self-contradicting in many places.  This PR fixes this and also improves formatting and wording a bit.

When the document was contradicting itself, I picked the version that I personally prefer. Please let me know in case you disagree with my choices.

## Remaining Inconsistencies
The following inconsistencies still remain, as I am not sure how it should be done:

1. Function names C vs C++. The guide uses camelCase for C++ functions but under_score for C. Do we really want different rules here?
1. Ordering of includes uses complete different ordering in the first part of the section and in the following example. Which order should we use?

## Things for Discussion
There are a few things in the current guide to which I disagree but which I don't want to change without consensus. So I would like to have your opinion on if we should keep it like this or change it.

1. "Comment as less as possible and as much as needed. By using good naming it should be possible to reduce comments to a minimum." -- I personally disagree with this. For someone who is new to the code, more comments are usually always better. If no one objects, I would remove this paragraph.
1. Single line `if`s don't need braces. -- I would prefer to *always* enforce braces as this is a common source for bugs (when later a second line is added to the if but no braces).
1. "Short conditional statements may be written on one line if this enhances
readability" (example `if (x == kFoo) return new Foo();`). I don't think that this ever enhances readability. Also it is related to the point above (braces).

### CODE REVIEW

**Comment by jbck** [comment]
I 100% agree with your suggestions, but I do think there are cases in which a _lot_ of single-line conditionals would be easier to read without extra lines or braces. So in general, #2 and #3 are only perhaps overruled by cases in which there are an objective "many" such lines all one after another.

**Comment by jbck** [comment]
Regarding naming, I dunno. I think they should be the same, but I would like an opinion from others (@fjes and @JauriG).

**Comment by zivke** [comment]
1) Function names - I personally like the camel case names simply because it makes the distinction between variables and functions really easy.
2) Includes order - I personally prefer to list all of the project headers first (starting with the related header file), then other libraries, then C/C++ headers, but this is not that important as long as they are grouped in some way.
3) Comments - I think that these two sentences are the best way to say to someone to try to find the balance in comment length. I would leave them.
4) If/else - I am ALWAYS for brackets! I hate when people use if/else without them.
5) I'm against one-liners like these.

**Comment by fjes** [comment]
As for the naming I agree this should be consistent to underscore_names.  @zivke what is the difference between a pointer variable and a pointer variable pointing to a function? I don't see too much room for misunderstanding, variables usually don't have `()` appended. ;)

@zivke as for the include order, this matters in C (and in C++ I assume). Sometimes definitions are wilfully or accidentally overwritten. I strongly recommend to start with the local headers and make exceptions when necessary. But the exception has to be a mindful decision.

ad (1): People tend to comment _what_ and not _why_, so I wouldn't encourage a too wordy code documentation.
ad (2) + (3): I prefer a consistency in the if-statements, especially with the block-braces. It is simply too easy to mess this up. So yes, always braces, if-one-liners are not encouraged but in rare cases I could imagine a exception.

I also would like to allow exceptions if the exception has advantages in _special_ cases.

**Comment by jbck** [comment]
Seems like we're pretty close to an agreement. Let's discuss it in the next meeting so everybody else can get their input: https://www.wrike.com/open.htm?id=289084200

**Comment by luator** [comment]
@jbck @fjes @jfilba I applied all the remaining changes as discussed this morning.
